src/snappy_device_agents/devices/multi/tfclient.py

```diff
@@ -145,6 +147,10 @@ def cancel_job(self, job_id):
         """Tell the server to cancel a specified job_id"""
         try:
             self.post(f"/v1/job/{job_id}/action", {"action": "cancel"})
+        except requests.exceptions.HTTPError as exc:
+            # Ignore it if the job is already cancelled or completed
+            if exc.response.status_code != 400:
+                raise
```

Comment:
I wonder, could it be that the server returns 400 even if the job wasn't cancelled or completed? For instance perhaps just malformed input?

Reply:
According to the code, yes it could be: https://github.com/canonical/testflinger/blob/main/src/api/v1.py#L277-L293

Reply:
Good catch, @omar-selo and @nadzyah thanks! First off, yes it will also return a 400 for unsupported action. However, this is changing to 422 (Unprocessable) due to the schema and openapi spec library I'm using. It does some built-in validation of the data you receive, which is really nice, and I think 422 fits much nicer for that. So that part will be fixed in a different PR.

As for bad job id, how do you feel about using 404 - "Resource not found"? I think I've seen this not just for bad URLs, but also for situations where the resource you're specifying as part of the URL is missing (ex. foo.com/users/42 if user 42 doesn't exist). If so, then there might be some other places we want to change to 404 for consistency.

Reply:
Both make sense to me. I was also wondering whether it's reasonable to return 200 from backend if the job was already canceled.